erts/emulator/beam/erl_map.c

```diff
@@ -3501,6 +3501,58 @@ BIF_RETTYPE erts_internal_map_next_3(BIF_ALIST_3) {
         BIF_ERROR(BIF_P, BADARG);
     }
 
+    /* Handle an ordered iterator. */
+    if (type == iterator && (is_list(path) || is_nil(path))) {
+#ifdef DEBUG
+#define ORDERED_ITER_FACTOR 200
+#else
+#define ORDERED_ITER_FACTOR 32
+#endif
+        int orig_elems = MAX(1, ERTS_BIF_REDS_LEFT(BIF_P) / ORDERED_ITER_FACTOR);
+        int elems = orig_elems;
+        Uint needed = 4 * elems + 2;
+        Eterm *hp = HAlloc(BIF_P, needed);
+        Eterm *hp_end = hp + needed;
+        Eterm result = am_none;
+        Eterm *patch_ptr = &result;
+
+        while (is_list(path) && elems > 0) {
+            Eterm *lst = list_val(path);
+            Eterm key = CAR(lst);
+            Eterm res = make_tuple(hp);
+            const Eterm *value = erts_maps_get(key, map);
+            if (!value) {
+            ordered_badarg:
+                HRelease(BIF_P, hp_end, hp);
+                BIF_ERROR(BIF_P, BADARG);
+            }
+            hp[0] = make_arityval(3);
+            hp[1] = key;
+            hp[2] = *value;
+            *patch_ptr = res;
+            patch_ptr = &hp[3];
+            hp += 4;
+            path = CDR(lst);
+            elems--;
+        }
+
+        if (is_list(path)) {
+            Eterm next = CONS(hp, path, map);
+            hp += 2;
+            *patch_ptr = next;
+            BUMP_ALL_REDS(BIF_P);
+            ERTS_ASSERT(result == am_none || is_tuple(result));
```

Comment:
I don't see how `result` could be `am_none` here.

and it should probably be an ASSERT macro for debug build only.

Reply:
and I think you could also ASSERT(hp == hp_end).

Reply:
Yes, it should be `ASSERT()`, and `result` can only be `am_none` if the list is empty. Fixed.